zio-http/src/main/scala/zhttp/http/middleware/Metrics.scala

```diff
@@ -0,0 +1,102 @@
+package zhttp.http.middleware
+
+import zhttp.http.{Http, Middleware, Request, Response}
+import zio._
+import zio.metrics.{Metric, MetricKeyType, MetricLabel}
+
+private[zhttp] trait Metrics {
+  def metrics[R, E](
+    mapper: Metrics.Mapper = Map.empty,
+    boundaries: MetricKeyType.Histogram.Boundaries = Metrics.defaultBoundaries,
+  ): HttpMiddleware[R with Metrics.InFlightRequests, E] =
+    new Middleware[R with Metrics.InFlightRequests, E, Request, Response, Request, Response] {
+      private def metricLabelsForRequest(req: Request): Set[MetricLabel] =
+        Set(
+          MetricLabel("method", req.method.toString),
+          MetricLabel("path", mapper.lift(req).getOrElse(req.path.toString())),
+        )
+
+      private def withRequestMeta[Type, In, Out](metric: Metric[Type, In, Out])(req: Request) =
+        metric.tagged(metricLabelsForRequest(req))
+
+      private def requestsTotal(req: Request) =
+        withRequestMeta(Metric.counterInt("http_requests_total"))(req)
+
+      private val errorCount = Metric.counterInt("http_errors_total")
+
+      private def errorCountWithRequest(req: Request) =
+        withRequestMeta(errorCount)(req)
+
+      private def errorCountWithResponse(req: Request, res: Response) =
+        errorCountWithStatus(req, res.status.code.toString)
+
+      private def errorCountWithStatus(req: Request, status: String) =
+        errorCountWithRequest(req).tagged("status_code", status)
+
+      private def requestDuration(req: Request) =
+        withRequestMeta(
+          Metric.histogram(
+            "http_request_duration_seconds",
+            boundaries,
+          ),
+        )(req).trackDurationWith(toSeconds)
+
+      private val NANOS                  = 1_000_000_000.0d
+      private def toSeconds(d: Duration) =
+        (d.getSeconds() * NANOS + d.getNano()) / NANOS
+
+      def apply[R1 <: R with Metrics.InFlightRequests, E1 >: E](
+        http: Http[R1, E1, Request, Response],
+      ): Http[R1, E1, Request, Response] =
+        Http.fromOptionFunction[Request] { req =>
+          val labels = metricLabelsForRequest(req)
+          Metrics.InFlightRequests.increment(labels) *>
+            (http(req)
+              .ensuring(requestsTotal(req).increment *> Metrics.InFlightRequests.decrement(labels))
+              .tapBoth(
+                err =>
+                  err.fold(errorCountWithStatus(req, "404").increment)(_ => errorCountWithStatus(req, "500").increment),
+                res => ZIO.when(res.status.code > 399)(errorCountWithResponse(req, res).increment),
+              )
+              .tapDefect(_ => errorCountWithStatus(req, "500").increment)) @@ requestDuration(req)
+        }
+    }
+}
+
+object Metrics {
+  type Mapper = PartialFunction[Request, String]
```

Comment:
Can be used to override mappings (i.e map certain paths like `/users/1` to `/users/:id`) to avoid having Prometheus label count explode. It would be nice if we could deal with placeholders automatically. Perhaps via the new API work @kitlangton is doing?